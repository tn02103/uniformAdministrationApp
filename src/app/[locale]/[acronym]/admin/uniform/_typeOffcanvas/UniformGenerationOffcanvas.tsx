import { LabelIconButton } from "@/components/Buttons/LabelIconButton";
import { InputFormField } from "@/components/fields/InputFormField";
import { SelectFormField } from "@/components/fields/SelectFormField";
import { ToggleFormField } from "@/components/fields/ToggleFormField";
import { useModal } from "@/components/modals/modalProvider";
import { createUniformGeneration, deleteUniformGeneration, updateUniformGeneration } from "@/dal/uniform/generation/_index";
import { useUniformSizelists, useUniformTypeList } from "@/dataFetcher/uniformAdmin";
import { useI18n } from "@/lib/locales/client";
import { SAFormHandler } from "@/lib/SAFormHandler";
import { UniformGeneration } from "@/types/globalUniformTypes";
import { uniformGenerationFormSchema, UniformGenerationFormType } from "@/zod/uniformConfig";
import { zodResolver } from "@hookform/resolvers/zod";
import { useEffect, useState } from "react";
import { Button, Col, Offcanvas, Row } from "react-bootstrap";
import { FormProvider, useForm } from "react-hook-form";
import { z } from "zod";


type UniformGenerationOffcanvasProps = {
    uniformTypeId: string;
    usingSizes: boolean;
    generation: UniformGeneration | null;
    onHide: () => void;
}
export const UniformgenerationOffcanvas = ({ generation, uniformTypeId, usingSizes, onHide }: UniformGenerationOffcanvasProps) => {
    const formSchema = uniformGenerationFormSchema.omit({ fk_sizelist: true }).extend({
        fk_sizelist: usingSizes
            ? z.string().min(1, 'string.required').uuid()
            : uniformGenerationFormSchema.shape.fk_sizelist,
    });

    const t = useI18n();
    const modal = useModal();
    const form = useForm<UniformGenerationFormType>({
        defaultValues: generation ?? { fk_sizelist: null },
        mode: "onTouched",
        resolver: zodResolver(formSchema),
    });

    const [editable, setEditable] = useState(!generation);
    const { sizelistList } = useUniformSizelists();
    const { mutate } = useUniformTypeList();
    const sizelistOptions = sizelistList?.map(sl => ({ value: sl.id, label: sl.name })) ?? [];

    useEffect(() => {
        form.reset(generation ?? undefined);
    }, [generation]);

    const handleDelete = () => {
        if (!generation) return;
        const deleteMutation = () => {
            mutate(deleteUniformGeneration(generation.id));
            onHide();
        }

        modal?.dangerConfirmationModal({
            header: t('admin.uniform.generationList.deleteModal.header', { generation: generation.name }),
            message: <span>
                {t('admin.uniform.generationList.deleteModal.message.part1')}<br />
                {t('admin.uniform.generationList.deleteModal.message.part2')}<br />
                {t('admin.uniform.generationList.deleteModal.message.part3')}<br />
            </span>,
            confirmationText: t('admin.uniform.generationList.deleteModal.confirmationText', { generation: generation.name }),
            dangerOption: {
                option: t('common.actions.delete'),
                function: deleteMutation,
            }
        });
    }

    const handleSave = (data: UniformGenerationFormType) => {
        if (!generation) {
            handleCreate(data);
            return;
        }

        SAFormHandler(
            updateUniformGeneration({ data, id: generation.id }),
            form.setError,
            (data) => { mutate(data); setEditable(false); },
            t('common.error.actions.save'),
        )
    }
    const handleCreate = (data: UniformGenerationFormType) => {
        SAFormHandler(
            createUniformGeneration({ ...data, uniformTypeId }),
            form.setError,
            (data) => { mutate(data); onHide(); },
            t('common.error.actions.create'),
        );
    }

    const handleCancel = () => {
        if (!generation) {
            onHide();
            return;
        }

        setEditable(false);
        form.reset(generation);
    }

    return (
        <Offcanvas show={true} onHide={onHide} placement="end" backdrop={false} style={{ width: "500px" }}>
            <Offcanvas.Header closeButton>
                <Offcanvas.Title>
                    <h2>{generation ? generation.name : t('admin.uniform.generationList.header.create')}</h2>
                </Offcanvas.Title>
            </Offcanvas.Header>
            <Offcanvas.Body>
                <hr className="my-0" />
                <Row className="mb-4 justify-content-evenly">
                    <LabelIconButton
                        variantKey="edit"
                        disabled={editable}
                        onClick={() => setEditable(true)}
                    />
                    <LabelIconButton
                        variantKey="delete"
                        disabled={editable}
                        onClick={() => handleDelete()}
                    />
                </Row>
                <FormProvider {...form}>
                    <form onSubmit={form.handleSubmit(handleSave)}>
                        <Row>
                            <Col xs={6}>
                                <InputFormField<UniformGenerationFormType>
                                    name="name"
                                    label={t('common.name')}
                                    plaintext={!editable}
                                    disabled={!editable}
                                />
                            </Col>
                            <Col xs={6} className="align-bottom">
                                <ToggleFormField<UniformGenerationFormType>
                                    name="outdated"
                                    label={t('common.uniform.generation.outdated')}
                                    disabled={!editable}
                                />
                            </Col>
                            {usingSizes && (
                                <Col xs={6}>
                                    <SelectFormField<UniformGenerationFormType>
                                        name="fk_sizelist"
                                        label={t('common.uniform.sizelist.label')}
                                        options={sizelistOptions}
                                        plaintext={!editable}
                                        disabled={!editable}
                                        labelClassName={(generation && usingSizes && (!generation?.fk_sizelist)) ? "text-danger" : ""}
                                    />
                                </Col>
                            )}
                            {editable && (
                                <Row className="justify-content-evenly mt-2 mb-4">
                                    <Button className="col-auto" type="submit" >{t(generation ? 'common.actions.save' : 'common.actions.create')}</Button>
                                    <Button className="col-auto" variant="outline-secondary" onClick={handleCancel}>{t('common.actions.cancel')}</Button>
                                </Row>
                            )}
                        </Row>
                    </form>
                </FormProvider>
            </Offcanvas.Body>
        </Offcanvas>
    );
}