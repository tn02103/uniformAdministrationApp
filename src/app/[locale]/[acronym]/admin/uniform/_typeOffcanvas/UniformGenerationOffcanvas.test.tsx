import { UniformgenerationOffcanvas } from "@/app/[locale]/[acronym]/admin/uniform/_typeOffcanvas/UniformGenerationOffcanvas";
import { render, screen } from "@testing-library/react";
import user from "@testing-library/user-event";


const sizeListIds = [
    'e667d674-7df8-436b-a2b8-77b06e063d36',
    'a961545b-28a7-409e-9200-1d85ccd53522',
    '07de1d59-4fc6-447b-98a6-da916e5792ef',
]
const uniformTypeId = '2b255769-2154-479f-bfca-c6a49e523334';
const testGeneration = {
    id: "testId",
    name: "Test Generation",
    outdated: false,
    fk_sizelist: sizeListIds[0],
    sortOrder: 0,
    sizelist: {
        id: sizeListIds[0],
        name: "Test Size List",
    },
}


jest.mock("@/dataFetcher/uniformAdmin", () => {
    const typeListMutate = jest.fn();
    return {
        useUniformSizelists: jest.fn(() => ({
            sizelistList: [{ id: sizeListIds[0], name: "Test Size List" }, { id: sizeListIds[1], name: "Test Size List 2" }, { id: sizeListIds[2], name: "Test Size List 3" }],
        })),
        useUniformTypeList: jest.fn(() => ({
            mutate: typeListMutate,
        })),
    };
});
jest.mock("@/lib/locales/client", () => {
    return {
        useI18n: jest.fn(() => (key: string) => key),
    };
});
jest.mock("@/components/modals/modalProvider", () => {
    const dangerModal = jest.fn();
    return {
        useModal: jest.fn(() => ({
            dangerConfirmationModal: dangerModal,
        })),
    };
});
jest.mock("@/dal/uniform/generation/_index", () => {
    return {
        createUniformGeneration: jest.fn(async () => "uniform generation created"),
        deleteUniformGeneration: jest.fn(async () => "uniform generation deleted"),
        updateUniformGeneration: jest.fn(async () => "uniform generation updated"),
    };
});
jest.mock('@/lib/locales/client', () => {
    return ({
        useScopedI18n: jest.fn().mockImplementation((scope: string) => {
            return function (key: string, values?: any) {
                return `${scope}.${key}`;
            }
        }),
        useI18n: jest.fn().mockImplementation(() => jest.fn((key) => key)),
        useCurrentLocale: jest.fn(() => ({
            locale: "de",
            setLocale: jest.fn(),
        })),
    });
});
jest.mock("@/components/errorMessage", () => {
    return function ErrorMessage({ error, ariaLabel, ...divProps}: { error: string, testId: string, ariaLabel: string }) {
        return <div className="text-danger fs-7" role="alert" aria-label={ariaLabel} {...divProps}>{error}</div>;
    };
});

Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
    })),
});

describe('<UniformgenerationOffcanvas>', () => {
    const { updateUniformGeneration, createUniformGeneration, deleteUniformGeneration } = require('@/dal/uniform/generation/_index');
    const { useUniformTypeList, useUniformSizelists } = require('@/dataFetcher/uniformAdmin');
    const { useModal } = require('@/components/modals/modalProvider');

    afterEach(() => jest.clearAllMocks());

    describe('with generation', () => {
        it('should render the component', async () => {
            render(
                <UniformgenerationOffcanvas
                    generation={testGeneration}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={() => { }}
                />
            );

            const sizeSelect = screen.queryByText('common.uniform.sizelist.label');
            expect(sizeSelect).toBeInTheDocument();


            const nameInput = screen.getByRole('textbox', { name: 'common.name' });
            const sizeSelectDiv = screen.getByRole('paragraph', { name: 'common.uniform.sizelist.label' });
            const outdatedSwitch = screen.getByRole('switch', { name: 'common.uniform.generation.outdated' });

            expect(nameInput).toHaveValue(testGeneration.name);
            expect(sizeSelectDiv).toHaveTextContent(testGeneration.sizelist.name);
            expect(outdatedSwitch).toHaveAttribute('aria-checked', String(testGeneration.outdated));

            expect(screen.getByRole('dialog')).toMatchSnapshot();
        });
        it('hide sizelist if usingSizes is false', async () => {
            render(
                <UniformgenerationOffcanvas
                    generation={testGeneration}
                    uniformTypeId={uniformTypeId}
                    usingSizes={false}
                    onHide={() => { }}
                />
            );

            const sizeSelect = screen.queryByText('common.uniform.sizelist.label');
            expect(sizeSelect).not.toBeInTheDocument();
        });
        it('should mark sizelist-label if usingSizes but sizelist is null', async () => {
            render(
                <UniformgenerationOffcanvas
                    generation={{ ...testGeneration, fk_sizelist: null }}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={() => { }}
                />
            );

            const sizeSelect = screen.queryByText('common.uniform.sizelist.label');
            expect(sizeSelect).toBeInTheDocument();
            expect(sizeSelect).toHaveClass('text-danger');
        });
        it('should set editable state', async () => {
            render(
                <UniformgenerationOffcanvas
                    generation={testGeneration}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={() => { }}
                />
            );

            const editButton = screen.getByText('common.actions.edit');
            const deleteButton = screen.getByText('common.actions.delete');
            const nameInput = screen.getByRole('textbox', { name: 'common.name' });
            const sizeSelectDiv = screen.getByRole('paragraph', { name: 'common.uniform.sizelist.label' });
            const outdatedSwitch = screen.getByRole('switch', { name: 'common.uniform.generation.outdated' });

            let saveButton = screen.queryByText('common.actions.save');
            let cancelButton = screen.queryByText('common.actions.cancel');
            let sizeSelect = screen.queryByRole('combobox', { name: 'common.uniform.sizelist.label' });

            expect(nameInput).toBeVisible();
            expect(nameInput).toHaveClass('form-control-plaintext');
            expect(nameInput).toHaveAttribute('disabled');
            expect(outdatedSwitch).toBeVisible();
            expect(outdatedSwitch).toHaveAttribute('aria-disabled', 'true');
            expect(sizeSelect).toBeNull();
            expect(sizeSelectDiv).toBeVisible();


            expect(editButton).toBeVisible();
            expect(deleteButton).toBeVisible();
            expect(saveButton).toBeNull();
            expect(cancelButton).toBeNull();

            expect(editButton).not.toHaveAttribute('disabled');
            expect(deleteButton).not.toHaveAttribute('disabled');

            await user.click(editButton);
            saveButton = screen.queryByText('common.actions.save');
            cancelButton = screen.queryByText('common.actions.cancel');
            sizeSelect = screen.getByRole('combobox', { name: 'common.uniform.sizelist.label' });

            expect(nameInput).toHaveClass('form-control');
            expect(nameInput).not.toHaveAttribute('disabled');
            expect(outdatedSwitch).toHaveAttribute('aria-disabled', 'false');
            expect(sizeSelect).toBeVisible();
            expect(sizeSelect).not.toHaveAttribute('disabled');
            expect(sizeSelectDiv).not.toBeVisible();

            expect(editButton).toBeVisible();
            expect(deleteButton).toBeVisible();
            expect(saveButton).toBeVisible();
            expect(cancelButton).toBeVisible();
            expect(editButton).toHaveAttribute('disabled');
            expect(deleteButton).toHaveAttribute('disabled');
        });

        it('resets on cancel', async () => {
            render(
                <UniformgenerationOffcanvas
                    generation={testGeneration}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={() => { }}
                />
            );

            const editButton = screen.getByText('common.actions.edit');
            await user.click(editButton);


            const nameInput = screen.getByRole('textbox', { name: 'common.name' });
            const outdatedSwitch = screen.getByRole('switch', { name: 'common.uniform.generation.outdated' });
            const sizeSelect = screen.getByRole('combobox', { name: 'common.uniform.sizelist.label' });
            const cancelButton = screen.getByText('common.actions.cancel');

            await user.clear(nameInput);
            await user.type(nameInput, 'New Name');
            await user.click(outdatedSwitch);
            await user.selectOptions(sizeSelect, sizeListIds[1]);
            expect(nameInput).toHaveValue('New Name');
            expect(outdatedSwitch).toHaveAttribute('aria-checked', 'true');
            expect(sizeSelect).toHaveValue(sizeListIds[1]);
            await user.click(cancelButton);

            expect(nameInput).toHaveValue(testGeneration.name);
            expect(outdatedSwitch).toHaveAttribute('aria-checked', String(testGeneration.outdated));
            expect(sizeSelect).not.toBeInTheDocument();
            const sizeSelectDiv = screen.getByRole('paragraph', { name: 'common.uniform.sizelist.label' });
            expect(sizeSelectDiv).toHaveTextContent(testGeneration.sizelist.name);
        });
    });
    describe('without generation', () => {
        it('should render in editable state if no generation is passed', async () => {
            render(
                <UniformgenerationOffcanvas
                    generation={null}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={() => { }}
                />
            );

            expect(screen.getByRole('dialog')).toMatchSnapshot();
        });

        it('should hide on Cancel', async () => {
            const onHide = jest.fn();
            render(
                <UniformgenerationOffcanvas
                    generation={null}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={onHide}
                />
            );

            const cancelButton = screen.getByText('common.actions.cancel');
            await user.click(cancelButton);
            expect(onHide).toHaveBeenCalledTimes(1);
            expect(onHide).toHaveBeenCalledWith();
        });
    });
    describe('dal methods', () => {
        it('should delete generation', async () => {
            deleteUniformGeneration.mockReturnValue('uniform generation deleted');
            const onHide = jest.fn();

            let dangerOption;
            const { dangerConfirmationModal } = useModal();
            dangerConfirmationModal.mockImplementation((option: any) => dangerOption = option.dangerOption);
            const { mutate } = useUniformTypeList();

            render(
                <UniformgenerationOffcanvas
                    generation={testGeneration}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={onHide}
                />
            );

            const deleteButton = screen.getByText('common.actions.delete');
            await user.click(deleteButton);

            expect(dangerConfirmationModal).toHaveBeenCalled();
            expect(dangerConfirmationModal).toHaveBeenCalledWith({
                header: 'admin.uniform.generationList.deleteModal.header',
                message: expect.anything(),
                confirmationText: 'admin.uniform.generationList.deleteModal.confirmationText',
                dangerOption: {
                    option: 'common.actions.delete',
                    function: expect.any(Function),
                },
            });
            expect(dangerOption).toBeDefined();

            await dangerOption!.function();
            expect(deleteUniformGeneration).toHaveBeenCalledTimes(1);
            expect(deleteUniformGeneration).toHaveBeenCalledWith(testGeneration.id);
            expect(mutate).toHaveBeenCalledTimes(1);
            expect(mutate).toHaveBeenCalledWith("uniform generation deleted");
            expect(onHide).toHaveBeenCalledTimes(1);
        });
        it('should call save function', async () => {
            const onHide = jest.fn();
            const { mutate } = useUniformTypeList();

            render(
                <UniformgenerationOffcanvas
                    generation={testGeneration}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={onHide}
                />
            );

            const editButton = screen.getByText('common.actions.edit');
            await user.click(editButton);

            const nameInput = screen.getByRole('textbox', { name: 'common.name' });
            const outdatedSwitch = screen.getByRole('switch', { name: 'common.uniform.generation.outdated' });
            const sizeSelect = screen.getByRole('combobox', { name: 'common.uniform.sizelist.label' });
            const saveButton = screen.getByText('common.actions.save');

            await user.clear(nameInput);
            await user.type(nameInput, 'New Name');
            await user.click(outdatedSwitch);
            await user.selectOptions(sizeSelect, sizeListIds[1]);

            expect(nameInput).toHaveValue('New Name');
            expect(outdatedSwitch).toHaveAttribute('aria-checked', 'true');
            expect(sizeSelect).toHaveValue(sizeListIds[1]);

            await user.click(saveButton);
            expect(updateUniformGeneration).toHaveBeenCalledTimes(1);
            expect(updateUniformGeneration).toHaveBeenCalledWith({
                data: {
                    name: 'New Name',
                    outdated: true,
                    fk_sizelist: sizeListIds[1],
                },
                id: testGeneration.id,
            });

            expect(mutate).toHaveBeenCalledTimes(1);
            expect(mutate).toHaveBeenCalledWith("uniform generation updated");
            expect(onHide).toHaveBeenCalledTimes(0);
        });
        it('should catch update form-errors', async () => {
            updateUniformGeneration.mockImplementationOnce(async () => {
                return {
                    error: {
                        message: "custom.uniform.generation.nameDuplication",
                        formElement: "name",
                    }
                };
            });

            render(
                <UniformgenerationOffcanvas
                    generation={testGeneration}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={() => { }}
                />
            );

            const editButton = screen.getByText('common.actions.edit');
            await user.click(editButton);

            const nameInput = screen.getByRole('textbox', { name: 'common.name' });
            const saveButton = screen.getByText('common.actions.save');

            await user.clear(nameInput);
            await user.type(nameInput, 'New Name');
            await user.click(saveButton);

            expect(updateUniformGeneration).toHaveBeenCalledTimes(1);
            expect(updateUniformGeneration).toHaveBeenCalledWith({
                data: {
                    name: 'New Name',
                    outdated: false,
                    fk_sizelist: sizeListIds[0],
                },
                id: testGeneration.id,
            });

            const errorMessage = await screen.findByText('custom.uniform.generation.nameDuplication');
            expect(errorMessage).toBeInTheDocument();
        });
        it('should call create function', async () => {
            const onHide = jest.fn();
            const { mutate } = useUniformTypeList();

            render(
                <UniformgenerationOffcanvas
                    generation={null}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={onHide}
                />
            );

            const nameInput = screen.getByRole('textbox', { name: 'common.name' });
            const outdatedSwitch = screen.getByRole('switch', { name: 'common.uniform.generation.outdated' });
            const sizeSelect = screen.getByRole('combobox', { name: 'common.uniform.sizelist.label' });
            const createButton = screen.getByText('common.actions.create');

            await user.type(nameInput, 'New Name');
            await user.click(outdatedSwitch);
            await user.selectOptions(sizeSelect, sizeListIds[1]);
            expect(nameInput).toHaveValue('New Name');
            expect(outdatedSwitch).toHaveAttribute('aria-checked', 'true');
            expect(sizeSelect).toHaveValue(sizeListIds[1]);
            await user.click(createButton);

            expect(createUniformGeneration).toHaveBeenCalledTimes(1);
            expect(createUniformGeneration).toHaveBeenCalledWith({
                name: 'New Name',
                outdated: true,
                fk_sizelist: sizeListIds[1],
                uniformTypeId: uniformTypeId,
            });

            expect(mutate).toHaveBeenCalledTimes(1);
            expect(mutate).toHaveBeenCalledWith("uniform generation created");
            onHide();
        });
        it('should catch create form-errors', async () => {
            createUniformGeneration.mockImplementationOnce(async () => {
                return {
                    error: {
                        message: "custom.uniform.generation.nameDuplication",
                        formElement: "name",
                    }
                };
            });

            render(
                <UniformgenerationOffcanvas
                    generation={null}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={() => { }}
                />
            );

            const nameInput = screen.getByRole('textbox', { name: 'common.name' });
            const sizeSelect = screen.getByRole('combobox', { name: 'common.uniform.sizelist.label' });
            const createButton = screen.getByText('common.actions.create');

            await user.type(nameInput, 'New Name');
            await user.selectOptions(sizeSelect, sizeListIds[0]);
            await user.click(createButton);

            expect(createUniformGeneration).toHaveBeenCalledTimes(1);
            expect(createUniformGeneration).toHaveBeenCalledWith({
                name: 'New Name',
                outdated: false,
                fk_sizelist: sizeListIds[0],
                uniformTypeId: uniformTypeId,
            });

            const errorMessage = await screen.findByText('custom.uniform.generation.nameDuplication');
            expect(errorMessage).toBeInTheDocument();
        });
    });
    describe('form validation', () => {
        it('should show error if name is empty', async () => {
            render(
                <UniformgenerationOffcanvas
                    generation={testGeneration}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={() => { }}
                />
            );

            const editButton = screen.getByText('common.actions.edit');
            await user.click(editButton);

            const nameInput = screen.getByRole('textbox', { name: 'common.name' });
            const saveButton = screen.getByText('common.actions.save');

            await user.clear(nameInput);
            await user.click(saveButton);

            expect(updateUniformGeneration).toHaveBeenCalledTimes(0);

            const errorMessage = await screen.findByRole('alert', { name: 'error message name' });
            expect(errorMessage).toBeVisible();
            expect(errorMessage).toHaveTextContent('string.required');
        });

        it('should show error if sizelist is empty and usingSizes', async () => {
            render(
                <UniformgenerationOffcanvas
                    generation={null}
                    uniformTypeId={uniformTypeId}
                    usingSizes={true}
                    onHide={() => { }}
                />
            );

            const nameInput = screen.getByRole('textbox', { name: 'common.name' });
            const createButton = screen.getByText('common.actions.create');
            await user.type(nameInput, 'New Name');
            await user.click(createButton);

            const sizelistError = screen.getByRole('alert', {name: 'error message fk_sizelist'});
            expect(sizelistError).toBeVisible();
            expect(sizelistError).toHaveTextContent('pleaseSelect');
        });
        it('should allow sizelist empty if not usingSizes', async () => {
            render(
                <UniformgenerationOffcanvas
                    generation={null}
                    uniformTypeId={uniformTypeId}
                    usingSizes={false}
                    onHide={() => { }}
                />
            );

            const nameInput = screen.getByRole('textbox', { name: 'common.name' });
            const createButton = screen.getByText('common.actions.create');
            await user.type(nameInput, 'New Name');
            await user.click(createButton);

            expect(createUniformGeneration).toHaveBeenCalledTimes(1);
        });
    });
});
