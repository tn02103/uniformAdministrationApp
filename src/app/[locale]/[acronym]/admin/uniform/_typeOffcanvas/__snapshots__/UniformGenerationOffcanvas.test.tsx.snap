// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UniformgenerationOffcanvas> with generation should render the component 1`] = `
<div
  aria-modal="true"
  class="offcanvas offcanvas-end offcanvas-toggling show"
  role="dialog"
  style="width: 500px; visibility: visible;"
  tabindex="-1"
>
  <div
    class="offcanvas-header"
  >
    <div
      class="offcanvas-title h5"
    >
      <h2>
        Test Generation
      </h2>
    </div>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    />
  </div>
  <div
    class="offcanvas-body"
  >
    <hr
      class="my-0"
    />
    <div
      class="mb-4 justify-content-evenly row"
    >
      <button
        class="col-auto border-0 btn btn-outline-secondary"
        type="button"
      >
        common.actions.edit
         
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-pen fa-sm "
          data-icon="pen"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M362.7 19.3L314.3 67.7 444.3 197.7l48.4-48.4c25-25 25-65.5 0-90.5L453.3 19.3c-25-25-65.5-25-90.5 0zm-71 71L58.6 323.5c-10.4 10.4-18 23.3-22.2 37.4L1 481.2C-1.5 489.7 .8 498.8 7 505s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L421.7 220.3 291.7 90.3z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="col-auto border-0 btn btn-outline-danger"
        type="button"
      >
        common.actions.delete
         
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-trash fa-sm "
          data-icon="trash"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M135.2 17.7L128 32 32 32C14.3 32 0 46.3 0 64S14.3 96 32 96l384 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-96 0-7.2-14.3C307.4 6.8 296.3 0 284.2 0L163.8 0c-12.1 0-23.2 6.8-28.6 17.7zM416 128L32 128 53.2 467c1.6 25.3 22.6 45 47.9 45l245.8 0c25.3 0 46.3-19.7 47.9-45L416 128z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <form>
      <div
        class="row"
      >
        <div
          class="col-6"
        >
          <div
            class="mb-3"
          >
            <label
              class="fw-bold m-0 form-label"
              for="input-name"
            >
              common.name
            </label>
            <input
              aria-invalid="false"
              class="form-control-plaintext"
              disabled=""
              id="input-name"
              name="name"
              type="text"
              value="Test Generation"
              width="auto"
            />
            <div
              aria-label="error message name"
              class="text-danger fs-7"
              id="err_name"
              role="alert"
              testid="err_name"
            />
          </div>
        </div>
        <div
          class="align-bottom col-6"
        >
          <div
            class="mb-3"
          >
            <label
              class="fw-bold m-0 form-label"
              for="toggle-outdated"
            >
              common.uniform.generation.outdated
            </label>
            <div
              class="form-check form-switch"
            >
              <input
                aria-checked="false"
                aria-disabled="true"
                class="form-check-input fw-bold"
                id="toggle-outdated"
                name="outdated"
                role="switch"
                style="pointer-events: none; opacity: 0.5;"
                tabindex="-1"
                type="checkbox"
                value="false"
              />
            </div>
            <div
              class="text-danger fs-7"
              role="alert"
              testid="err_outdated"
            />
          </div>
        </div>
        <div
          class="col-6"
        >
          <div
            class="mb-3"
          >
            <label
              class="fw-bold m-0  form-label"
              for="select-fk_sizelist"
            >
              common.uniform.sizelist.label
            </label>
            <p
              aria-label="common.uniform.sizelist.label"
              aria-readonly="true"
              class="py-2 m-0"
            >
              Test Size List
            </p>
            <div
              aria-label="error message fk_sizelist"
              class="text-danger fs-7"
              id="err_fk_sizelist"
              role="alert"
              testid="err_fk_sizelist"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`<UniformgenerationOffcanvas> without generation should render in editable state if no generation is passed 1`] = `
<div
  aria-modal="true"
  class="offcanvas offcanvas-end offcanvas-toggling show"
  role="dialog"
  style="width: 500px; visibility: visible;"
  tabindex="-1"
>
  <div
    class="offcanvas-header"
  >
    <div
      class="offcanvas-title h5"
    >
      <h2>
        admin.uniform.generationList.header.create
      </h2>
    </div>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    />
  </div>
  <div
    class="offcanvas-body"
  >
    <hr
      class="my-0"
    />
    <div
      class="mb-4 justify-content-evenly row"
    >
      <button
        class="col-auto border-0 btn btn-outline-secondary"
        disabled=""
        type="button"
      >
        common.actions.edit
         
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-pen fa-sm "
          data-icon="pen"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M362.7 19.3L314.3 67.7 444.3 197.7l48.4-48.4c25-25 25-65.5 0-90.5L453.3 19.3c-25-25-65.5-25-90.5 0zm-71 71L58.6 323.5c-10.4 10.4-18 23.3-22.2 37.4L1 481.2C-1.5 489.7 .8 498.8 7 505s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L421.7 220.3 291.7 90.3z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="col-auto border-0 btn btn-outline-danger"
        disabled=""
        type="button"
      >
        common.actions.delete
         
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-trash fa-sm "
          data-icon="trash"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M135.2 17.7L128 32 32 32C14.3 32 0 46.3 0 64S14.3 96 32 96l384 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-96 0-7.2-14.3C307.4 6.8 296.3 0 284.2 0L163.8 0c-12.1 0-23.2 6.8-28.6 17.7zM416 128L32 128 53.2 467c1.6 25.3 22.6 45 47.9 45l245.8 0c25.3 0 46.3-19.7 47.9-45L416 128z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <form>
      <div
        class="row"
      >
        <div
          class="col-6"
        >
          <div
            class="mb-3"
          >
            <label
              class="fw-bold m-0 form-label"
              for="input-name"
            >
              common.name
            </label>
            <input
              aria-invalid="false"
              class="form-control"
              id="input-name"
              name="name"
              type="text"
              value=""
              width="auto"
            />
            <div
              aria-label="error message name"
              class="text-danger fs-7"
              id="err_name"
              role="alert"
              testid="err_name"
            />
          </div>
        </div>
        <div
          class="align-bottom col-6"
        >
          <div
            class="mb-3"
          >
            <label
              class="fw-bold m-0 form-label"
              for="toggle-outdated"
            >
              common.uniform.generation.outdated
            </label>
            <div
              class="form-check form-switch"
            >
              <input
                aria-checked="false"
                aria-disabled="false"
                class="form-check-input fw-bold"
                id="toggle-outdated"
                name="outdated"
                role="switch"
                style="cursor: pointer;"
                tabindex="0"
                type="checkbox"
                value="false"
              />
            </div>
            <div
              class="text-danger fs-7"
              role="alert"
              testid="err_outdated"
            />
          </div>
        </div>
        <div
          class="col-6"
        >
          <div
            class="mb-3"
          >
            <label
              class="fw-bold m-0  form-label"
              for="select-fk_sizelist"
            >
              common.uniform.sizelist.label
            </label>
            <select
              aria-invalid="false"
              class="form-select"
              id="select-fk_sizelist"
              name="fk_sizelist"
            >
              <option
                disabled=""
              >
                common.error.pleaseSelect
              </option>
              <option
                value="e667d674-7df8-436b-a2b8-77b06e063d36"
              >
                Test Size List
              </option>
              <option
                value="a961545b-28a7-409e-9200-1d85ccd53522"
              >
                Test Size List 2
              </option>
              <option
                value="07de1d59-4fc6-447b-98a6-da916e5792ef"
              >
                Test Size List 3
              </option>
            </select>
            <div
              aria-label="error message fk_sizelist"
              class="text-danger fs-7"
              id="err_fk_sizelist"
              role="alert"
              testid="err_fk_sizelist"
            />
          </div>
        </div>
        <div
          class="justify-content-evenly mt-2 mb-4 row"
        >
          <button
            class="col-auto btn btn-primary"
            type="submit"
          >
            common.actions.create
          </button>
          <button
            class="col-auto btn btn-outline-secondary"
            type="button"
          >
            common.actions.cancel
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
